Secret and secret manager Tutorial (#2510)
harness/developer-hub: tutorials/platform/secrets-management.md

```

In this tutorial, we'll walk you through the process of safely storing and accessing sensitive information like API keys and credentials. You'll learn best practices while gaining hands-on experience to ensure your secrets remain protected. Let's dive into the world of Harness secrets management.

:::info
[Sign up today to unleash the potential of intelligent Harness CD](https://app.harness.io/auth/#/signup/?module=cd&utm_source=website&utm_medium=harness-developer-hub&utm_campaign=cd-plg&utm_content=tutorials-cd-secret).
:::

## Create secrets

Harness offers built-in secret management for encrypted storage of sensitive information. Secrets are decrypted when needed, and only the private network-connected Harness Delegate has access to the key management system. You can also integrate your own secret manager. To learn more about secrets in Harness, go to [Harness Secret Manager Overview](https://developer.harness.io/docs/platform/Secrets/Secrets-Management/harness-secret-manager-overview/).

To create a secret, do the following:

1. Sign in to Harness.
2. Select **Projects**, and then select **Secrets**.
3. Select **New Secret**, and then select **Text**.

> 4. Enter the secret name `SECRET_NAME`.

Reviewer: ```suggestion
4. Enter  **Name** for the secret.
```

> 5. For the **Secret Value**, paste the personal access token (PAT) from GitHub with the repo scope. For more information, go to [creating a personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) in the GitHub documentation. 

Reviewer: ```suggestion
5. For the **Secret Value**, enter a GitHub personal access token (PAT) with the repo scope. For more information, go to [creating a personal access token](https://help.github.com/en/github/authenticating-to-github/creating-a-personal-access-token-for-the-command-line) in the GitHub documentation. 
```

1. Select **New Secret**, and then select **Text**.
2. Enter the secret name `aws_kms_access_key`.
3. For the secret value, paste the access token for your AWS user account. The Harness Delegate uses this credential to authenticate Harness with AWS at deployment runtime.
4. Select **Save**.

### Create an AWS secret key ID

To create an AWS secret key ID, do the following:

1. Select **New Secret**, and then select **Text**.
2. Enter the secret name `aws_kms_secret_key`.
3. For the secret value, paste the access token for your AWS user account. The Harness Delegate uses this credential to authenticate Harness with AWS at deployment runtime.
4. Select **Save**.

### Create an AWS ARN

To create an AWS ARN, do the following:

1. Select **New Secret**, and then select **Text**.
2. Enter the secret name `aws_kms_arn`.
3. For the secret value, paste the ARN for your AWS user account. The Harness Delegate uses this credential to authenticate Harness with AWS at deployment runtime.
4. Select **Save**.

For more information, go to [find the Access key ID and ARN](https://docs.aws.amazon.com/kms/latest/developerguide/find-cmk-id-arn.html) in the AWS documentation.

> ## Connectors

Reviewer: ```suggestion
### Connectors
```

should this be level 3?

1. Create the **Hashicorp Connector**.

2. Copy the contents of [hashicorp_vault_connector.yaml](https://github.com/harness-community/harnesscd-example-apps/blob/master/harness-platform/secrets/hashicorp_vault_connector.yaml).
3. In your Harness project in the Harness Manager, under **Project Setup**, select **Connectors**.
4. Select **Create via YAML Builder** and paste the copied YAML.
5. Replace the `Public Vault URL` with your public vault URL from Hashicorp Vault.
6. Select **Save Changes** and verify that the new connector named **hashicorp_vault** is successfully created.
7. Select **Connection Test** under **Connectivity Status** to verify the connection is successful.

:::info note
If you are NOT using the default organization and project, make sure to change the `orgIdentifier` and `projectIdentifier` in the `hashicorp_vault_connector.yaml` file.
:::

You can now use this secret manager in your pipeline.

> ---------------------------------------------------

Reviewer: This line is only used on one of the tabs

Verify that you have the following:

- The **AWS - Access Key ID** from AWS
- The **AWS - Secret Key ID** from AWS
- A fork of the [harnessed-example-apps](https://github.com/harness-community/harnesscd-example-apps/fork) repository through the GitHub website.
    - For more information on forking a GitHub repository, go to [forking a repository](https://docs.github.com/en/get-started/quickstart/fork-a-repo#forking-a-repository) in the GitHub documentation.
- **Docker**. For this tutorial, ensure that you have the Docker runtime installed on your Harness Delegate host. If you do not have the Docker runtime, use one of the following installation options:
    - [Docker for Mac](https://docs.docker.com/desktop/install/mac-install/)
    - [Docker for CentOS](https://docs.docker.com/engine/install/centos/)
    - [Docker for Ubuntu](https://docs.docker.com/engine/install/ubuntu/)
    - [Docker for Debian](https://docs.docker.com/engine/install/debian/)
    - [Docker for Windows](https://docs.docker.com/desktop/install/windows-install/) 
    - Check [Delegate system requirements](https://developer.harness.io/docs/platform/Delegates/delegate-concepts/delegate-requirements).

> ### Use AWS KMS to manage secrets 

Reviewer: ```suggestion
## Use AWS KMS to manage secrets 
```

should this heading be different? It's the same as on the AWS KMS tab.

Let's create a pipeline that will print logs on the console to see how Harness sanitization logs.

### Create a new secret

To create a new secret, do the following: 

1. Select **New Secret**, and then select **Text**.
2. Enter the secret name `docsecret`.
3. For the secret value, add the string `docsecret`.
4. Select **Save**.

### Create a pipeline

To create a pipeline, do the following:

1. In **Default Project**, select **Pipelines**.
2. Select **New Pipeline**.
3. Enter the name `print secret`.
4. Select **Inline** to store the pipeline in Harness.
5. Select **Start** and, in the Pipeline Studio, toggle to **YAML** to use the YAML editor.
6. Select **Edit YAML** to enable edit mode. Copy the contents of the[secret-sanitization.yaml file](https://github.com/harness-community/harnesscd-example-apps/blob/master/harness-platform/secrets/secret-sanitization.yaml) and paste it.
7. Select **Save** to save the pipeline.

### Run the pipeline

Finally, it's time to execute the pipeline.

To run the pipeline, do the following:

1. Select **Run**, and then select **Run Pipeline** to initiate the deployment.
2. Observe the execution logs as Harness runs the pipeline.

> 3. In console logs, you can see that the secret is masked with ***** as per below screenshot.

Reviewer: ```suggestion
3. In console logs, the secret is masked with asterisks (`*`), as shown in the following image.
```

<docimage path={require('./static/secret/secret-log.png')} width="60%" height="60%" title="Click to view full size image" />

> When a text secret is displayed in a deployment log, Harness substitutes the text secret value with asterisks (*) so that the secret value is never displayed.​ For more information, go to [secrets and log-sanitization](https://developer.harness.io/docs/platform/Secrets/Secrets-Management/secrets-and-log-sanitization).

Reviewer: ```suggestion
When a text secret is displayed in a deployment log, Harness substitutes the text secret value with asterisks (`*`) so that the secret value is never displayed.​ For more information, go to [secrets and log-sanitization](https://developer.harness.io/docs/platform/Secrets/Secrets-Management/secrets-and-log-sanitization).
```

Reviewer: FYI there is an exception that secrets exported in Output Variables are visible in build logs. https://developer.harness.io/docs/continuous-integration/use-ci/run-ci-scripts/run-step-settings#output-variables
